### PR TITLE
virt-launcher: Remove part of the domain event errors

### PR DESCRIPTION
**Pull request by oshoval**
Remove part of the domain event errors

Not all the domain errors need a server side
status refresh, which also, in turn cause
grpc server restart due to how its implemented.

In case of libvirt reconnection, (for example due to libvirtd restart) we do need the state
to be refreshed as soon as possible and this was left untouched.

The other cases were removed, which will improve stability and availability
the the gRPC server in the virt-handler, that serves all the VMs in the node.

Found during investigation of #3575

---

Add domain name to libvirt reconnect error

In order to be able to connect between the error,
and the vm that was triggering it, in the virt-handler side.

Signed-off-by: Or Shoval <oshoval@redhat.com>

```release-note
None
```

### CODE REVIEW

File: pkg/virt-handler/notify-server/server.go [hunk 1]
```diff
@@ -84,7 +84,7 @@ func (n *Notify) HandleDomainEvent(ctx context.Context, request *notifyv1.Domain
 	case string(watch.Deleted):
 		n.EventChan <- watch.Event{Type: watch.Deleted, Object: domain}
 	case string(watch.Error):
-		n.EventChan <- watch.Event{Type: watch.Error, Object: status}
```
Comment (davidvossel) [inline pkg/virt-handler/notify-server/server.go:87]: Would it make sense to consider a client side change instead to reduce errors to things we actually care about instead?

For example, if a libvirt event connection is reset, then that's a legitimate error the server should be aware of. But all these other things we're sending errors for, like "domain not found", probably shouldn't have been an error event to begin with.

@rmohr we need to prioritize your domain notify replacement at some point. I'm realizing more and more all the issues doing this serverside in virt-handler involves.

**Comment by davidvossel** [comment]
@oshoval 

I was poking in this area just now and I think i found a pretty big issue. [1] 

I introduced logic that pipes a socket in virt-launcher pod to the domain notify socket called domain-notify-pipe.sock. The issue is, if the domain socket is reset (which definitely happens when one of these Errors come in) then the domain-notify-pipe is broken.  We never re-estabilish that part of the connection. meaning notifies for all existing vmis on that host stop indefinitely. 

Basically, the dial in [1] needs to go into the for loop of [2] or else we're hosed when one of these error events come in. 


1. https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L214
2. https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L222

File: pkg/virt-handler/notify-server/server.go (hunk 1 above)
Comment (EdDev) [reply pkg/virt-handler/notify-server/server.go:87]: I am bit confused by the comment here.
Why would the "server" reset all connections with all virt-launchers due to a single client?
If I understand correctly, the only action taken by the virt-handler in such a case, is to reset the connections and nothing else. If the virt-handler wants to do something else, I guess a mechanism needs to be placed that changes the existing anyway.

So are you suggesting the current behavior is good or that we need to react to errors from the client side?
This change, AFAIU only makes sure we do not reset all connection if a client sent an error. Is this part ok?

**Comment by davidvossel** [comment]
I made #3818 to handle the reconnect issue. I still think the notion of reducing the Errors sent over this channel makes a lot of sense. So we should continue with this PR as well.

File: pkg/virt-handler/notify-server/server.go (hunk 1 above)
Comment (oshoval) [reply pkg/virt-handler/notify-server/server.go:87]: Yep if i understand right, there is no need to reset the gRPC medium which serves all the client, as Roman said here for example:
https://github.com/kubevirt/kubevirt/pull/3713#issuecomment-652261724

**Comment by oshoval** [comment]
> @oshoval
> 
> I was poking in this area just now and I think i found a pretty big issue. [1]
> 
> I introduced logic that pipes a socket in virt-launcher pod to the domain notify socket called domain-notify-pipe.sock. The issue is, if the domain socket is reset (which definitely happens when one of these Errors come in) then the domain-notify-pipe is broken. We never re-estabilish that part of the connection. meaning notifies for all existing vmis on that host stop indefinitely.
> 
> Basically, the dial in [1] needs to go into the for loop of [2] or else we're hosed when one of these error events come in.
> 
> 1. https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L214
> 2. https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L222

Thanks, great one

File: pkg/virt-handler/notify-server/server.go (hunk 1 above)
Comment (davidvossel) [reply pkg/virt-handler/notify-server/server.go:87]: the error makes the informer re-list/watch, pulling the latest state.

if you don't process errors, then the informer does not pull the latest state. This means if a libvirt reconnect occurs, the informer might not be accurate if anything changed during the time the connection to libvirt is being re-established.

Now, there's zero reason to recreate the grpc server when an error comes in. that's just a result of how the domain notify list/watch is implemented.

Comment (oshoval) [reply pkg/virt-handler/notify-server/server.go:87]: Thanks, even that, we need just for some error cases or for all?
will learn if its safe to remove the Stop, in case the watcher will use the same interface again (according a comment there it doesnt, but i will try to check)

Comment (EdDev) [reply pkg/virt-handler/notify-server/server.go:87]: > the error makes the informer re-list/watch, pulling the latest state.
> 
> if you don't process errors, then the informer does not pull the latest state. This means if a libvirt reconnect occurs, the informer might not be accurate if anything changed during the time the connection to libvirt is being re-established.
> 
> Now, there's zero reason to recreate the grpc server when an error comes in. that's just a result of how the domain notify list/watch is implemented.

Are you saying here that we should not ignore the errors coming in but have an action of refreshing the state?
If this is the case, then this PR merits a nack.
Could you please clarify this?

Comment (oshoval) [reply pkg/virt-handler/notify-server/server.go:87]: nack in the current state, a little improvement would fix it according what i understood (and according 
https://github.com/kubevirt/kubevirt/pull/3806#issuecomment-657806368)

you can remove the lgtm already if you want, thanks

**Comment by EdDev** [comment]
Ignoring the error may not be correct.
/lgtm cancel

File: pkg/virt-handler/notify-server/server.go (hunk 1 above)
Comment (oshoval) [reply pkg/virt-handler/notify-server/server.go:87]: Updated PR, please take another look

**Comment by oshoval** [comment]
@davidvossel 
I see in a test 
i think i saw it before as well, (after the first take to fix it by updating grpc version), 
does #3830 handles it ?
I dont see there domain event ERROR (which this PR changes behavior of)
```
{"component":"virt-handler","level":"error","msg":"domain pipe listener closed.","pos":"vm.go:225","reason":"accept unix /proc/17216/root/var/run/kubevirt/domain-notify-pipe.sock: use of closed network connection","timestamp":"2020-07-15T16:51:30.575588Z"}
{"component":"virt-handler","kind":"","level":"info","msg":"closing notify pipe connection for vmi: read unix @-\u003e/var/run/kubevirt/domain-notify.sock: use of closed network connection","name":"testvmiw5xzhm6nbxlm58fsr9c6hlrqvngt62l5wxdmtcvlxhhcvmwm","namespace":"kubevirt-test-default","pos":"vm.go:245","timestamp":"2020-07-15T16:51:30.575674Z","uid":"d451568b-c804-4411-a66f-f729b1a3a12b"}
```

**Comment by oshoval** [comment]
@rmohr @davidvossel 
Would you please take a look ?

File: pkg/virt-launcher/notify-client/client.go [hunk 2]
```diff
@@ -201,7 +201,6 @@ func eventCallback(c cli.Connection, domain *api.Domain, libvirtEvent libvirtEve
 	if err != nil {
 		if !domainerrors.IsNotFound(err) {
 			log.Log.Reason(err).Error("Could not fetch the Domain.")
-			client.SendDomainEvent(newWatchEventError(err))
```
Comment (rmohr) [inline pkg/virt-launcher/notify-client/client.go:204]: I think one thing to consider is, if libvirtd gets e.g. killed an restarted, if we will now still see an update of the latest domain state being transferred over to virt-handler. Is that the case?

Comment (oshoval) [reply pkg/virt-launcher/notify-client/client.go:204]: i think this one is ?
https://github.com/kubevirt/kubevirt/pull/3806/files#diff-7175e08324a35a860b12d6518d396988L339
and if so, it will be the case that ERROR would be sent, as David said (hopefully that i understood right)

File: pkg/virt-launcher/notify-client/client.go [hunk 3]
```diff
@@ -337,7 +334,7 @@ func (n *Notifier) StartDomainNotifier(
 				eventCallback(domainConn, domainCache, libvirtEvent{}, n, deleteNotificationSent,
 					interfaceStatuses, guestOsInfo)
 			case <-reconnectChan:
-				n.SendDomainEvent(newWatchEventError(fmt.Errorf("Libvirt reconnect")))
+				n.SendDomainEvent(newWatchEventError(fmt.Errorf("Libvirt reconnect, domain %s", domainName)))
```
Comment (davidvossel) [inline pkg/virt-launcher/notify-client/client.go:337]: looks great!
